### ISSUE TITLE
Ubuntu 20.04 Root on ZFS - No mirrored ESP Partition

### ISSUE THREAD

**Issue by JohannesNeu** [openzfs/openzfs-docs#107]
@rlaager, I have the following issue with the Ubuntu 20.04 Root on ZFS HOWTO:

Hi,
I am at step 6.5 "dpkg-reconfigure grub-efi-amd64". I can run the command and questions will be asked, but there is no possibility to choose the EFI partitions from a list. I have a mirror setup with two SSDs and each has an EFI partition, a BOOT partition and a ROOT partition. 
These are the questions I am asked:
![grafik](https://user-images.githubusercontent.com/39328632/104819961-b350f080-5831-11eb-81ea-0d59a0e991b8.png)
![grafik](https://user-images.githubusercontent.com/39328632/104819966-c19f0c80-5831-11eb-854b-281de2b9960a.png)
![grafik](https://user-images.githubusercontent.com/39328632/104819970-cebbfb80-5831-11eb-8573-1013af796d9c.png)

And this is the result:
![grafik](https://user-images.githubusercontent.com/39328632/104820003-19d60e80-5832-11eb-8103-ad1e276e62c5.png)

Is it possible that this is connected to my /etc/fstab setup?
![grafik](https://user-images.githubusercontent.com/39328632/104820018-4558f900-5832-11eb-87d3-93d22d5a5298.png)

Thanks for your help.

**Comment by rlaager** [comment]
I'm not sure. I've certainly seen that be an issue before the system reboots (which is why I save that step until after the reboot). I'm not sure about the /etc/fstab entry, but you could try making it `/dev/disk/by-uuid/B948-62E2 /boot/efi vfat defaults 0 0` instead. If that works, let me know and I'll have to think about removing the recently added (c00aa93389260573349303e95cb3e6e3cb805b09) automount idea I guess?

Are you running into this issue:
https://github.com/openzfs/openzfs-docs/issues/95

**Comment by JohannesNeu** [comment]
@rlaager Thanks for the link, I totally missed that issue.
I checked: and it seems I also have grubenv:


`ich@host:/boot/grub$ ls -la`
`total 1644`
`drwxr-xr-x 4 root root       7 Jan 17 09:37 .`
`drwxr-xr-x 4 root root      12 Jan 16 16:02 ..`
`drwxr-xr-x 2 root root       3 Jan 16 14:00 fonts`
`-r--r--r-- 1 root root   13791 Jan 17 09:37 grub.cfg`
`-rw-r--r-- 1 root root    1024 Jan 17 09:27 grubenv`
`-rw-r--r-- 1 root root 2395475 Jan 16 18:34 unicode.pf2`
`drwxr-xr-x 2 root root     279 Jan 16 14:00 x86_64-efi`

should I just remove it and rerun dpkg reconfigure?

**Comment by rlaager** [comment]
Yeah, give that a shot and let me know if that works. I'll need to dig into this more for a proper fix.

**Comment by JohannesNeu** [comment]
@rlaager deleting grubenv did not help. I still get the same error as before (I did sudo dpkg-reconfigure grub-efi-amd64)

**Comment by JohannesNeu** [comment]
I installed some new packages and everytime I use apt I get a grub related error. I can still reboot but maybe this helps to find out what is happening?

```
Setting up grub-efi-amd64-signed (1.142.10+2.04-1ubuntu26.8) ...
Trying to migrate /boot/efi into esp config
dpkg: error processing package grub-efi-amd64-signed (--configure):
 installed grub-efi-amd64-signed package post-installation script subprocess returned error exit status 1
Processing triggers for man-db (2.9.1-1) ...
dpkg: dependency problems prevent processing triggers for shim-signed:
 shim-signed depends on grub-efi-amd64-signed | grub-efi-arm64-signed; however:
  Package grub-efi-amd64-signed is not configured yet.
  Package grub-efi-arm64-signed is not installed.

dpkg: error processing package shim-signed (--configure):
 dependency problems - leaving triggers unprocessed
Processing triggers for systemd (245.4-4ubuntu3.4) ...
Errors were encountered while processing:
 grub-efi-amd64-signed
 shim-signed
ZSys is adding automatic system snapshot to GRUB menu
E: Sub-process /usr/bin/dpkg returned an error code (1)
```

**Comment by rlaager** [comment]
The "Trying to migrate" code is in `/usr/lib/grub/grub-multi-install`. I don't know what's breaking, but here's what I would try next:

Edit `/usr/lib/grub/grub-multi-install`. Find this block:
```
# We either migrate /boot/efi over, or we check if we have invalid devices
if [ -z "$RET" ] && [ "$seen" != "true" ]; then
  echo "Trying to migrate /boot/efi into esp config"
  esp="$(get_mounted_device /boot/efi)"
  if [ "$esp" ]; then
    esp="$(device_to_id "$esp")"
  fi
  if [ "$esp" ]; then
    db_set grub-efi/install_devices "$esp"
    db_fset grub-efi/install_devices seen true
    RET="$esp"
  fi
```

Add some debug code like this:
```
# We either migrate /boot/efi over, or we check if we have invalid devices
if [ -z "$RET" ] && [ "$seen" != "true" ]; then
  echo "Trying to migrate /boot/efi into esp config"
  echo "HERE1" >> /tmp/grub-debug.log
  esp="$(get_mounted_device /boot/efi)"
  echo "HERE2: $esp" >> /tmp/grub-debug.log
  if [ "$esp" ]; then
    esp="$(device_to_id "$esp")"
    echo "HERE3 $esp" >> /tmp/grub-debug.log
  fi
  if [ "$esp" ]; then
    echo "HERE4" >> /tmp/grub-debug.log
    db_set grub-efi/install_devices "$esp"
    db_fset grub-efi/install_devices seen true
    RET="$esp"
  fi
  echo "HERE5" >> /tmp/grub-debug.log
```

Make the package reconfigure, e.g. `apt install -f` or `dpkg-reconfigure grub-efi-amd64-signed`. The goal is to get the same "Trying to migrate" message. Once you ran that code, look at `/tmp/grub-debug.log` and paste it here. That should help narrow down at which step it is failing. I'm guessing it is either failing in `get_mounted_device` or `device_to_id`. Once we know where it is failing and the value involved, then we can dig into that code to try to figure out why.

**Comment by JohannesNeu** [comment]
Hi @rlaager 
I added the code and and did apt-get dist-upgrade which caused:

```
Setting up grub-efi-amd64-signed (1.142.10+2.04-1ubuntu26.8) ...
Trying to migrate /boot/efi into esp config
dpkg: error processing package grub-efi-amd64-signed (--configure):
 installed grub-efi-amd64-signed package post-installation script subprocess returned error exit status 1
```

The resulting entries in /tmp/grub-debug.log are:
```
HERE1
HERE2: systemd-1
```

**Comment by rlaager** [comment]
That `systemd-1` is really weird looking. That is coming from `get_mounted_device /boot/efi` which parses `/proc/mounts`. Can you paste (or attach) the contents of `/proc/mounts`?

**Comment by JohannesNeu** [comment]
I checked:
```
admin@host:~$ cat /proc/mounts 
sysfs /sys sysfs rw,nosuid,nodev,noexec,relatime 0 0
proc /proc proc rw,nosuid,nodev,noexec,relatime 0 0
udev /dev devtmpfs rw,nosuid,noexec,relatime,size=8141532k,nr_inodes=2035383,mode=755 0 0
devpts /dev/pts devpts rw,nosuid,noexec,relatime,gid=5,mode=620,ptmxmode=000 0 0
tmpfs /run tmpfs rw,nosuid,nodev,noexec,relatime,size=1636672k,mode=755 0 0
rpool/ROOT/ubuntu_azjs5b / zfs rw,relatime,xattr,posixacl 0 0
securityfs /sys/kernel/security securityfs rw,nosuid,nodev,noexec,relatime 0 0
tmpfs /dev/shm tmpfs rw,nosuid,nodev 0 0
tmpfs /run/lock tmpfs rw,nosuid,nodev,noexec,relatime,size=5120k 0 0
tmpfs /sys/fs/cgroup tmpfs ro,nosuid,nodev,noexec,mode=755 0 0
cgroup2 /sys/fs/cgroup/unified cgroup2 rw,nosuid,nodev,noexec,relatime 0 0
cgroup /sys/fs/cgroup/systemd cgroup rw,nosuid,nodev,noexec,relatime,xattr,name=systemd 0 0
pstore /sys/fs/pstore pstore rw,nosuid,nodev,noexec,relatime 0 0
efivarfs /sys/firmware/efi/efivars efivarfs rw,nosuid,nodev,noexec,relatime 0 0
none /sys/fs/bpf bpf rw,nosuid,nodev,noexec,relatime,mode=700 0 0
cgroup /sys/fs/cgroup/freezer cgroup rw,nosuid,nodev,noexec,relatime,freezer 0 0
cgroup /sys/fs/cgroup/net_cls,net_prio cgroup rw,nosuid,nodev,noexec,relatime,net_cls,net_prio 0 0
cgroup /sys/fs/cgroup/cpuset cgroup rw,nosuid,nodev,noexec,relatime,cpuset,clone_children 0 0
cgroup /sys/fs/cgroup/hugetlb cgroup rw,nosuid,nodev,noexec,relatime,hugetlb 0 0
cgroup /sys/fs/cgroup/cpu,cpuacct cgroup rw,nosuid,nodev,noexec,relatime,cpu,cpuacct 0 0
cgroup /sys/fs/cgroup/perf_event cgroup rw,nosuid,nodev,noexec,relatime,perf_event 0 0
cgroup /sys/fs/cgroup/rdma cgroup rw,nosuid,nodev,noexec,relatime,rdma 0 0
cgroup /sys/fs/cgroup/memory cgroup rw,nosuid,nodev,noexec,relatime,memory 0 0
cgroup /sys/fs/cgroup/pids cgroup rw,nosuid,nodev,noexec,relatime,pids 0 0
cgroup /sys/fs/cgroup/blkio cgroup rw,nosuid,nodev,noexec,relatime,blkio 0 0
cgroup /sys/fs/cgroup/devices cgroup rw,nosuid,nodev,noexec,relatime,devices 0 0
systemd-1 /proc/sys/fs/binfmt_misc autofs rw,relatime,fd=28,pgrp=1,timeout=0,minproto=5,maxproto=5,direct,pipe_ino=21536 0 0
hugetlbfs /dev/hugepages hugetlbfs rw,relatime,pagesize=2M 0 0
mqueue /dev/mqueue mqueue rw,nosuid,nodev,noexec,relatime 0 0
debugfs /sys/kernel/debug debugfs rw,nosuid,nodev,noexec,relatime 0 0
tracefs /sys/kernel/tracing tracefs rw,nosuid,nodev,noexec,relatime 0 0
fusectl /sys/fs/fuse/connections fusectl rw,nosuid,nodev,noexec,relatime 0 0
configfs /sys/kernel/config configfs rw,nosuid,nodev,noexec,relatime 0 0
rpool/lxdmount/datenbank-var /zmnt/lxddaten/datenbank/var zfs rw,relatime,xattr,posixacl 0 0
rpool/USERDATA/admin_fllsy /home/admin zfs rw,relatime,xattr,posixacl 0 0
rpool/USERDATA/root_azjs5b /root zfs rw,relatime,xattr,posixacl 0 0
/dev/loop1 /snap/lxd/19032 squashfs ro,nodev,relatime 0 0
/dev/loop0 /snap/core18/1944 squashfs ro,nodev,relatime 0 0
/dev/loop2 /snap/snapd/10707 squashfs ro,nodev,relatime 0 0
bpool/BOOT/ubuntu_azjs5b /boot zfs rw,nodev,relatime,xattr,posixacl 0 0
systemd-1 /boot/efi autofs rw,relatime,fd=49,pgrp=1,timeout=60,minproto=5,maxproto=5,direct,pipe_ino=32824 0 0
bpool/grub /boot/grub zfs rw,nodev,relatime,xattr,posixacl 0 0
rpool/lxdmount/anzeige-var /zmnt/lxddaten/anzeige/var zfs rw,relatime,xattr,posixacl 0 0
tmpfs /run/snapd/ns tmpfs rw,nosuid,nodev,noexec,relatime,size=1636672k,mode=755 0 0
tmpfs /var/snap/lxd/common/ns tmpfs rw,relatime,size=1024k,mode=700 0 0
nsfs /var/snap/lxd/common/ns/shmounts nsfs rw 0 0
nsfs /var/snap/lxd/common/ns/mntns nsfs rw 0 0
binfmt_misc /proc/sys/fs/binfmt_misc binfmt_misc rw,nosuid,nodev,noexec,relatime 0 0
/dev/loop3 /snap/lxd/19188 squashfs ro,nodev,relatime 0 0
tmpfs /run/user/1000 tmpfs rw,nosuid,nodev,relatime,size=1636668k,mode=700,uid=1000,gid=1000 0 0
/dev/loop4 /snap/core18/1988 squashfs ro,nodev,relatime 0 0
/dev/loop5 /snap/snapd/11036 squashfs ro,nodev,relatime 0 0
nsfs /run/snapd/ns/lxd.mnt nsfs rw 0 0
```

**Comment by rlaager** [comment]
From what I can see, you chose the new automounting `/boot/efi` alternative and that's breaking GRUB.

Apparently I didn't test that well enough. I can't remember if I tried a full install for that or just changed it on a working system. Sorry!

I've just removed that option from the HOWTO. If you convert to having a "regular" entry for `/boot/efi` in `/etc/fstab`, I suspect everything will work fine. If not, let me know.

**Comment by antonc42** [comment]
@rlaager I am having a similar problem. I followed the directions at https://openzfs.github.io/openzfs-docs/Getting%20Started/Ubuntu/Ubuntu%2020.04%20Root%20on%20ZFS.html on 2021-07-08. I am installing on 4 drives in a raidz configuration. Everything appears to work fine other than this `dpkg-reconfigure grub-efi-amd64` issue. It doesn't give me an option to select the drives as the original poster said.

I am not having any issues with `grub-efi-amd64-signed`.

I tried removing the `/boot/grub/grubenv` file and re-running `dpkg-reconfigure grub-efi-amd64`, but it did not help. I experimented to see what would happen if I removed the first drive and it did not boot.

I then unmounted the first disk from `/boot/efi` and mounted the second disk's first partition at `/boot/efi` and ran `dpkg-reconfigure grub-efi-amd64`. Then when I removed the first drive and rebooted, the system came up. I'm not sure if this is really a viable solution though. I would probably have to manually do this on every update, correct?

**Comment by AndrewPaglusch** [comment]
I wanted to confirm what @antonc42 said in his latest update. After switching my disks from virtio to SATA emulation, the `dpkg-reconfigure grub-efi-amd64` step worked as expected and I was able to choose the EFI partitions.

**Comment by inthebrilliantblue** [comment]
> I wanted to confirm what @antonc42 said in his latest update. After switching my disks from virtio to SATA emulation, the `dpkg-reconfigure grub-efi-amd64` step worked as expected and I was able to choose the EFI partitions.

I just came across this problem in a VM for testing using virtio instead of sata disks. Could not get the dpkg-reconfigure to bring up efi disk selection until I changed the virtio disks to sata disks. After booting up, and running dpkg-reconfigure grub-efi-amd64, it would then show the two EFI partitions.

Im assuming that because the system is presenting the disks with /dev/vdX that the reconfigure doesnt show them in a selection, where as /dev/sdX disks do. I havent tried doing alias names to see if it would work on virtio yet, but wanted to confirm that this is an issue with virtio in VMs as well. I dont know if this is intended or a "wont fix" bug, but it is one that affected me in testing.
